### PR TITLE
Enable Github depend bot to scan for Javascript vulnerability

### PR DESCRIPTION
**Pull request by antonysigma**
Add a dummy javascript library dependencies file "package.json" and the book-keeping file "yarn.lock" to monitor the IRVisualizer tool for outdated dependencies. Activate Github's Dependbot notification for any security vulnerabilities and virus.

Upgrade the JQuery library to v3.7.0 .

cc'ed @maaz139 and @mcourteaux .

### CODE REVIEW

File: src/irvisualizer/README.md
Comment (mcourteaux) [inline src/irvisualizer/README.md:23]: typo

**Comment by mcourteaux** [comment]
The security check seems useful, but the extra build dependency on yarn seems very unpleasant. Is this mandatory, or only to update the `yarn.lock` file? What's the purpose of that file anyway? Doesn't Github check the dependencies through the `package.json`?

**Comment by antonysigma** [comment]
> Is this mandatory, or only to update the `yarn.lock` file? What's the purpose of that file anyway? Doesn't Github check the dependencies through the `package.json`?

Thank you for the feedback. Okay. Removing the `yarn.lock` file.

**Review by mcourteaux — CHANGES_REQUESTED** [review]

**Comment by mcourteaux** [comment]
@steven-johnson This PR was made concurrently with mine (#7755) and unintentionally reverts my changes (due to some refactoring and code moving from one file to another it seems). Please don't merge, until this is addressed.

**Comment by steven-johnson** [comment]
Thanks for the heads-up :-)

**Comment by mcourteaux** [comment]
@steven-johnson @antonysigma I am very sorry for the confusion. I saw some code in the file that I was confident that I deleted (the whole part about init_tooltips). But it seemed I just forgot to delete that code. I changed the tooltips to be CSS-only, which warrants deleting some javascript, which I forgot it seems. So not outdated! Apologies!

**Comment by mcourteaux** [comment]
I haven't tested this PR yet. I could try that tomorrow, it's late in Belgium right now. :)

**Comment by mcourteaux** [comment]
Okay gonna test now!

**Comment by mcourteaux** [comment]
Okay, I can confirm it works. I have no clue if this will actually make GitHub keep an eye on this, but at least it doesn't break anything as far as I can tell.

**Review by mcourteaux — APPROVED** [review]
~~LGMT~~ Minor comment: see below.

File: src/irvisualizer/README.md
Comment (steven-johnson) [inline src/irvisualizer/README.md:5]: s/Dependbot/Dependabot/

**Review by steven-johnson — COMMENTED** [review]
LGTM pending green buildbots

**Review by mcourteaux — CHANGES_REQUESTED** [review]
Why is there a symlink from the template.js to the main.js? If you just rename the file, Git will keep track of the commit history as it will treat it as a rename. Using a symlink breaks it.

Also, maybe the indent can be undone in main.js?

**Comment by mcourteaux** [comment]
@antonysigma I liked the .js extension you made. So putting the `<script>` tag around it and having HTML feels inferior to what you originally had. You can just rename this file from `*.html` to `*.js` and Git should understand.

Just to complete: right now, you also have a nested `<script>` tag, because it's once in the HTML file, and once in the C++ generator code.

File: src/irvisualizer/StmtToViz_javascript.template.html
Comment (mcourteaux) [inline src/irvisualizer/StmtToViz_javascript.template.html:1]: So, to be clear what I mean: if you rename this file to be `StmtToViz_javascript.template.js`, and remove the `<script>` tag, I think it'd be ideal.

**Review by mcourteaux — CHANGES_REQUESTED** [review]

File: src/irvisualizer/StmtToViz_javascript.template.html
Comment (antonysigma) [inline src/irvisualizer/StmtToViz_javascript.template.html:1]: > Just to complete: right now, you also have a nested <script> tag, because it's once in the HTML file, and once in the C++ generator code.

Tagging @mcourteaux to verify whether the nested `<script>` tags are introduced in the Halide build process.

Comment (mcourteaux) [reply src/irvisualizer/StmtToViz_javascript.template.html:1]: Sorry, was confused. Indeed, you deleted it, not added... Got the colors backwards, my bad.

File: src/irvisualizer/StmtToViz_javascript.template.html
Comment (antonysigma) [reply src/irvisualizer/StmtToViz_javascript.template.html:1]: I am afraid it will upset both the Makefile and CMake toolchain, and I don't know CMake enough to make the changes. I added your request to the Todo list in the Readme. Currently, my primary goal is to activate the Dependabot. 

May I rely on you to rename the template files without upsetting Makefile and CMake? Run `git grep html_template` and you will know what I mean. Also, the [bin2c](https://github.com/adobe/bin2c/blob/master/readme.md) is a nice reference; I guess the core-Halide developers reverse engineered `bin2c` to `binary2cpp` for Windows users.

**Comment by mcourteaux** [comment]
@antonysigma I see what you meant with this:

> It was an attempt to rename all templates files from `*.template.html` to `template/*.[js|css|html]`. Agreed it is a scope creep. Okay. I changed it back. The PR should be now simplier.

We __can__ turn this into a `*.js` file, if we add a build rule that does the binary2cpp for the JS file here:

https://github.com/halide/Halide/blob/acc9413084db07fd6c794e1d59b6eca0f72b140e/Makefile#L1190-L1191

and here:

https://github.com/halide/Halide/blob/acc9413084db07fd6c794e1d59b6eca0f72b140e/src/CMakeLists.txt#L362-L399

File: src/irvisualizer/StmtToViz_javascript.template.html
Comment (mcourteaux) [reply src/irvisualizer/StmtToViz_javascript.template.html:1]: Sure! I see. Let's keep it simple right now and keep the `.html` extension.

**Comment by mcourteaux** [comment]
To conclude this lenghty nitpicking discussion, let's keep the `<script>` tag right now, to make syntax highlighting work, because the `html` extension throws off the highlighters without the script-tag.

**Comment by mcourteaux** [comment]
@steven-johnson LGTM! I think this can land. :smile: :tada:

**Comment by antonysigma** [comment]
> let's keep the `<script>` tag right now, to make syntax highlighting work, because the `html` extension throws off the highlighters without the script-tag.

Done. Kindly asking @alexreinking to debug the CI build failure. I rebased the changes on top of `release_2013_11_11-8624-gacc941308`, not sure what went wrong.

**Comment by mcourteaux** [comment]
@antonysigma Thanks for taking the time here! I hope I wasn't too annoying :stuck_out_tongue_closed_eyes: Cool to see how the PR size went down over time. :smile:

File: src/irvisualizer/package.json
Comment (mcourteaux) [inline src/irvisualizer/package.json:12]: Oh, these still reference the other filename.

File: src/irvisualizer/package.json
Comment (mcourteaux) [inline src/irvisualizer/package.json:13]: Old filename.

File: src/irvisualizer/package.json
Comment (mcourteaux) [inline src/irvisualizer/package.json:14]: Old filename.

**Review by mcourteaux — CHANGES_REQUESTED** [review]

File: src/irvisualizer/package.json
Comment (antonysigma) [reply src/irvisualizer/package.json]: @mcourteaux Done. I highly doubt it will ever tun because `package.json` is a dummy file so far.

@steven-johnson is now resolving the CI issue. So I will refrain from `git push` till he has finished editing the PR.

Comment (steven-johnson) [reply src/irvisualizer/package.json]: This should be ready for you to resume editing, thanks for your patience.

Comment (antonysigma) [reply src/irvisualizer/package.json]: Thanks @steven-johnson . I finished editing. Closing this thread.

**Comment by steven-johnson** [comment]
This looks good, *except* we really must have a CMake-based build script for this as well; as you know, CMake is the reference build implementation for Halide, and we expect everything in it to be buildable via CMake. (Make support is provided on a best-effort basis by the contributors who care about it, but not all Halide features are support via Make, e.g. the Python bindings.)

**Comment by antonysigma** [comment]
Hi @steven-johnson , acknowledged the need to integrate all build process into Makefile and CMake. Since this PR's primary goal is to activate Github Dependabot, not to build anything, may I remove the Makefile to get this PR accepted?

The roadmap to trans-compile javascripts into `stmt_html` with Make/CMake integration is mentioned in the Readme.

**Comment by steven-johnson** [comment]
Let me see if I can put together a solution so we can land this

**Comment by steven-johnson** [comment]
OK, looking at this further, I don't think we need to bother with CMake -- TBH, the Makefile is so trivial even it could basically be moved into the README (but it's ok to leave it in).

That said, what I think *does* need to be added to the README is what tools are assumed to be installed on your system -- e.g. `yarn`, which may be present on all JavaScript expert's systems, but definitely wasn't present on mine (nor was Node, for that matter). Also `eslint`? etc etc

**Comment by steven-johnson** [comment]
ok, after doing `brew install node yarn eslint esbuild` on my Mac laptop, doing `make lint` results in:

```
$ make lint
yarn run lint
yarn run v1.22.19
$ eslint StmtToViz_javascript.template.html

/Users/srj/GitHub/Halide/src/irvisualizer/StmtToViz_javascript.template.html
  1:1  error  Parsing error: Unexpected token <

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
make: *** [lint] Error 1
```

**Comment by antonysigma** [comment]
> ok, after doing brew install node yarn eslint esbuild on my Mac laptop, doing make lint results in:

```
eslint StmtToViz_javascript.template.html

/Users/srj/GitHub/Halide/src/irvisualizer/StmtToViz_javascript.template.html
  1:1  error  Parsing error: Unexpected token <
```

Tagging @mcourteaux for the `package.json` roadmap discussion.

Steven has a question about the "redundant" `<script>` tag failure from the `eslint`. Eslint is the javascript static analyzer. He suggests we remove the `Makefile`, do you agree? The `eslint path/to/template/StmtToViz_main.js` idea with the `*.js` extension certainly can wait. I recalled you also have another PR to land asap, right? The `Makefile` and `eslint` idea can wait.

File: src/irvisualizer/README.md
Comment (antonysigma) [inline src/irvisualizer/README.md:28]: From @steven-johnson :

ok, after doing brew install node yarn eslint esbuild on my Mac laptop, doing make lint results in:

```
$ make lint
yarn run lint
yarn run v1.22.19
$ eslint StmtToViz_javascript.template.html

/Users/srj/GitHub/Halide/src/irvisualizer/StmtToViz_javascript.template.html
  1:1  error  Parsing error: Unexpected token <

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
make: *** [lint] Error 1
```

Yep, it is an TODO item: refactor the IRVisualizer code base to allow auto-formatting with `clang-format`, and then static analyzer checking with `eslint`. What you see with `Unexpected token <` is expected.

Comment (steven-johnson) [reply src/irvisualizer/README.md:28]: It's not at all clear that `make lint` is currently **expected** to fail -- I'd assume that all the build targets in the Makefile do what you expect. Please add appropriate comments and/or documentation -- I realize this may seem obvious to you but it certainly wasn't to me, so I suspect it may be non-obvious to others as well.

Comment (antonysigma) [reply src/irvisualizer/README.md:54]: Done moving the draft `Makefile` contents to Readme.md . Thank you for the guidance.

**Comment by mcourteaux** [comment]
> Tagging @mcourteaux for the `package.json` roadmap discussion.
> 
> Steven has a question about the "redundant" `<script>` tag failure from the `eslint` output. Eslint is the javascript static analyzer. He suggests we remove the `Makefile`, do you agree? The `eslint path/to/template/StmtToViz_main.js` idea with the `*.js` extension certainly can wait. I recalled you also have another PR #7793 to land asap, right? The `Makefile` and `eslint` idea can wait.

@antonysigma @steven-johnson  Right now, I still haven't run any of `eslint`, `yarn`, `node` or `esbuild`. My only requirement was that we do not need any of these tools to actually build Halide and the generators. Relying on any of these tools for in actual production use, would be a major turn-off for me. I was assuming these tools are nice-to-haves when actually working on the code of this HTML/JS stuff to get some more assistance, comfort of development, and static analysis reports.

So, to answer the question "agree to remove the Makefile?" Yes! The Makefile right now does not build anything it seems, but just _**describes**_ the tools you could use to get development assistance, if I interpret the situation correctly. To be honest, I don't see the benefit of a `make watch` over a `yarn run watch` (besides the automatic triggering of `yarn install`). This indeed should go in the README and probably in human readable format, instead of Makefile format. The benefit of a Makefile here seems minimal.

What _does_ seem useful is the static analysis (as most of us are used to compile-time validation of your code, instead of crashing at runtime), which could be automatically validated by the buildbots. But overall, this seems like a separate feature from the original idea of simply listing the dependencies to have an automatic scan from Dependabot.

On overall roadmap (potentially several bite-sized PRs) could be:

 1. Enable depandabot now by adding `package.json` (basically, the core bit of this PR).
 2. Rename the template files to have their proper extensions, e.g., `*.hmtl`, `*.js`, `*.css` and make sure all the Makefile and CMake build rules are up to date. Meanwhile, refactor the `<script>` tag to be generated by the C++ code.
 3. Accommodate an individual developer to run the JS/HTML specific linting tools, should they wish. This becomes more straightforward with the proper filetypes in place as per (2).
 4. Integrate linting as an *OPTIONAL* step in the build process in CMake and Makefile, and enable this in the buildbots.
 5. Once linting is an optional step in the build procedure (and the webtech tools are available on the build system), we can check out if we can produce these amalgamated offline builds of the HTML file which has the dependencies baked in.

Right now, this PR seems like it's grabbing bits of 1 and 3 without first solving 2. In the long term, I think reaching step 4 seems beneficial, as now PRs are mostly eyeballed and quickly tested to see if stuff still functions, but tools like this will probably improve the review process.

**Comment by mcourteaux** [comment]
To add to this, I'm willing to fix step 2, if this PR get trimmed to cover step 1. If we agree to proceed this way, then please do keep your work around because this `yarn` and `eslint` stuff seems useful, and you seem to know how it works (I don't... To be honest, I'm not sure what `yarn` is for, as I haven't needed it and everything did work. I do understand what the linting does tho :smile: ), so you or I could use it as reference material to proceed with step 3 of my proposed roadmap, once we have reached step 2.

**Comment by antonysigma** [comment]
Great! Deal! @steven-johnson I moved the Makefile contents to `README.md`. The PR is ready to land.

**Comment by steven-johnson** [comment]
At this point I will defer to @mcourteaux for approval.

**Comment by mcourteaux** [comment]
I was already getting annoyed by the way this JS stuff works. Was looking to get `yarn` to test the suggested commands in the README. Yarn seems to be a package manager, which you use to install the dependencies. Turns out, to install `yarn` you need `corepack` (some sort of module/package manager), that comes with `npm`, which is a package manager. :exploding_head: What? So I need package manager (apt) to install a package manager (npm), to install an alternative _experimental_ package manager `corepack`, to install another package manager `yarn`, to download three small scripts which are hosted online? All of this to replace a couple of `<script src="">` tags?

If this is the price we pay to get the generated HTML to be functioning offline, I'm not sure if this is the right path to take. To be honest, I'd much rather replace the jQuery, and bootstrap stuff with some simple Javascript and CSS. For the assembly syntax highlighting, there are two lightweight options:
 - it will then only work if you have internet, or the file is cached by your browser.
 - we just include this really small file in the repository (it's literally 2kB: https://cdn.jsdelivr.net/gh/speed-highlight/core/dist/index.js), which is possible, as the project is licensed CC0.

This seems like a reasonable compromise for a repository like Halide, as we'd stay out of the JS madhouse.

Any thoughts @steven-johnson @antonysigma? I hope I'm not being difficult, but I really feel like this stuff does not belong in a repo like Halide.

**Comment by steven-johnson** [comment]
> So I need package manager (apt) to install a package manager (npm), to install an alternative experimental package manager corepack, to install another package manager yarn, to download three small scripts which are hosted online? 

It's package managers all the way down

> If this is the price we pay to get the generated HTML to be functioning offline, I'm not sure if this is the right path to take.

I wish I could express useful thoughts about this, but my knowledge of "modern JS/HTML packaging" is, uh, ~20 years out of date -- I don't know what the current expectations are. (I do agree that needing to install multiple JS-specific tools seems like a surprising and suboptimal burden.)

> we just include this really small file in the repository (it's literally 2kB: https://cdn.jsdelivr.net/gh/speed-highlight/core/dist/index.js), which is possible, as the project is licensed CC0.

I *this* that CC0 would OK to include -- based on my reading of things it's considered "unencumbered", so as long as we add the right info to our LICENSE.txt we should be ok. That said: adding @abadams @jrk to see if they have opinions/concerns about this.

**Comment by abadams** [comment]
My javascript knowledge is also rudimentary at best, but I'll say I'm strongly in favor of minimizing the number of dependencies, and CC0 looks fine to me.

**Comment by antonysigma** [comment]
I can sense a subtle inclination to revert PR #7056 and #7421, so that `stmt_html` will remain as a syntax-highlighted, printable program listing. If that's the case, you all have my support, and let me retract the PR.

Hi @mcourteaux looks like we are not getting Dependabot with this PR. Please feel free to submit your own PR. The CVE concerns about the jQuery 1.x needs to be resolved. You are also welcome to take the bits and pieces of this PR to get your work done. Examples are: `.eslint.js` to flush out bugs in the Javascript; `.clang-format` to auto-format the javascript code.

**Comment by antonysigma** [comment]
Hi @abadams and @steven-johnson , I realized the PR to activate Github Dependabot, meant to reduce your review time, and to build confidence on the *Single Page Web Application (SPA)*, managed to destroy both. Sorry to diverting all the attention from other more critical PRs.

@mcourteaux 's concerns about the SPA and the lack of governance is legitimate. No matter I like it or not, with PRs #7056 and #7421, Halide's developers inadvertently embraced SPA built into the IR Visualizer design. `stmt_html` is no longer a pretty-printed program listing with syntax highlighting, but now becomes an SPA because it has hot path analysis, IR-to-assembly code seeking functions, as well as data movement + compute cost estimates in real time.

SPA is notorious for its lack of governance, and accelerated code rot. I roughly estimated 1.5 year decay half life of a typical SPA module. Once we embrace the SPA architecture, we will either need automation (e.g. Github Dependabot) or manual labor to keep it under control. Or, like @mcourteaux suggested, the Halide team can always steer back.

Similar projects encountered the "foreign/alien" SPA technology from the volunteers, and merged the PRs. But, they quickly grow wary of added complexity and extra maintenance. Each project eventually comes up with their own unique solutions. Feel free to reach out to them for advice. Examples are:

* Jupyter Notebook: https://github.com/jupyter/notebook/issues/6210#issuecomment-947830885

* Open hardware testing framework https://github.com/google/openhtf/tree/master/openhtf/output/web_gui https://github.com/google/openhtf/issues/1039

**Comment by mcourteaux** [comment]
I will attempt to do a major cleanup as I try to tackle #7519. Ideally, I'd like to kick out:

 - Bootstrap & bootstrap icons
 - jQuery
 - VizIR
 - Most of the JavaScript (without losing the functionality).

This should bring back the StmtHTML to a simple and clean page.

Additionally, I figured out that the split panels are very slow because of how the overflow of the content of the panels in the horizontal direction is handled. I believe I will be able to finally speed this aspect up as well.

**Comment by mcourteaux** [comment]
I believe @maaz139 might be on holidays or something. In the meanwhile, @steven-johnson do you know anything about the work on the `halide:darya-ver/ir-viz` branch? (Read last comment for more details.)

**Comment by abadams** [comment]
The commits in that branch also seem to be part of #7421, which is merged, so I wouldn't worry about it:

**Comment by abadams** [comment]
But looking at those commits, they have the same descriptions as the commits in #7421, so I think they're duplicates or something.

**Comment by maaz139** [comment]
Apologies for the slow response. I am currently *not* working on the visualizer although I am happy to help review / implement important changes. I'm not sure what happened here but the commits on #7421 have already been merged. I will close that branch.

**Comment by mcourteaux** [comment]
> Apologies for the slow response. I am currently _not_ working on the visualizer although I am happy to help review / implement important changes. I'm not sure what happened here but the commits on #7421 have already been merged. I will close that branch.

Okay! Cool. I'll give it a shot! It seems that @darya-ver has done a lot of work that went beyond the PR. You can see a demo of what she did here: https://darya-ver.github.io/files/project_files/visualize_halide_example_blur3x3.html and reflects the screenshot in #7056. So I'm not sure what's up with that.

**Pull request closed without merging**
